### DISCUSSION TITLE
[05.11.2024] Blog 4 - Software Design

### DISCUSSION THREAD

**Discussion by Kartoffeleintopf** [bunki1224/Softwareengineering-Projekt#28] [Show and tell]
## Einleitung
Unsere Reiseplaner-App benötigt verschiedene Klassen, um die verschiedenen Trips,  Events und Orte abzubilden. Hierfür haben wir Klassendiagramme entworfen, die zunächst die Hauptklassen wie Trip, Event, FlightEvent, HotelEvent, ActivityEvent und Location umfassen.
## Klassen Design
![UML Classes](https://github.com/user-attachments/assets/60b29103-e243-4fcc-86ae-9c7ad3388977)
Im großen gibt es 4 Klassen die wir als essentiell ansehen:
- Den User
- Den Trip
- Die Timeline
- Ein Event

Das Trip-Modell beispielsweise enthält neben seiner ID und Reiseziel (Destination) eine Timeline welche wiederum verschiedene Events speichert, die der Benutzer zu seiner Reise hinzufügen kann.
Events vererbt an verschiedene Typen von Events weiter, welche ihre eigenen Ausprägungen haben.
## SOLID-Prinzip
Ein wesentlicher Teil unseres Designs basiert auf dem Single Responsibility Principle (SRP), das besagt, dass jede Klasse eine einzige Verantwortlichkeit haben sollte. 
Beispielsweise wurde die Location-Klasse eigenständig gestaltet und kapselt alle Informationen zu einem geografischen Ort – Name, Koordinaten, Preis und Bewertung.
Auf ähnliche Weise sind auch FlightEvent, HotelEvent und ActivityEvent spezifische Unterklassen von Event, die jede für sich nur eine spezifische Art von Ereignis in der Timeline darstellen.
## Sequenzdiagramme
Unsere erstellten Sequenzdiagramme zeigen weiterhin Interaktion zwischen Komponenten und Modellen, und wir empfanden diese Abbildung als weiterhin akkurat. Daher wurden diese nicht aktualisiert und sind im [Blog 3](#27)  aufzufinden.

## Implementiere Klassen
Unsere implementierten Klassen sind in unserem Git Repo zu finden:
[Klassen](https://github.com/bunki1224/Softwareengineering-Projekt/tree/main/TripAhead/src/models)
<br>
> Euer TripAhead Team (CN)

**Comment by themsgcentre** [comment]
Vielen Dank für das Teilen eures Fortschritts – es ist spannend zu sehen, wie sich euer Projekt entwickelt! Euer Klassendiagramm stellt sehr gut die wichtigsten Komponenten eures Projekts dar und hilft dabei, die geplante Struktur besser nachzuvollziehen. Das Single Responsibility Prinzip ist dabei bemerkenswert gut eingehalten! Auch die Unterteilung in verschiedene Event-Typen mithilfe von Kindklassen zu realisieren, halten wir für eine effiziente Idee.

Uns würde noch interessieren, wie die verschiedenen Objekte miteinander interagieren; entsprechend wäre es gut, zusätzlich noch Assoziationen, Aggregationen und Kompositionen in das Diagramm einzufügen.

Weiter so – wir sind gespannt, wie das Projekt sich weiterentwickelt!

Viele Grüße,
Projekt TDC (HM)

**Comment by pplayer1712** [comment]
Ein sehr interessanter Beitrag zur Struktur der Reiseplaner-App! Die Orientierung an den SOLID-Prinzipien, insbesondere dem Single Responsibility Principle, ist gut durchdacht und macht das Design modular und flexibel. Die Trennung der verschiedenen Ereignistypen (wie Flight, Hotel, Activity) in eigene Klassen erleichtert es, spezifische Eigenschaften für jedes Event zu verwalten und bei Bedarf zu erweitern.

Die UML-Diagramme sind hilfreich, um die Beziehungen zwischen den Klassen zu verstehen – insbesondere die Vererbung von Events. Es wäre interessant, noch mehr über die Benutzererfahrung zu erfahren: Wie interagiert der User mit der Timeline und den Events? Insgesamt ein gelungenes Design für eine strukturierte und skalierbare App!

Viele Grüße
Projekt UMOC
/PL /JMS
